networkx/algorithms/coloring/greedy_coloring.py

```diff
@@ -65,14 +65,47 @@ def strategy_smallest_last(G, colors):
 
     """
     H = G.copy(with_data=False)
-
-    def min_and_pop():
-        """Removes and returns the minimum degree node in ``H``."""
-        node = min(H, key=H.degree)
-        H.remove_node(node)
-        return node
-
-    return reversed([min_and_pop() for x in G])
+    result = []
+
+    # Build initial degree list (i.e. the bucket queue data structure)
+    degrees = defaultdict(set)  # set(), for fast random-access removals
+    for node, d in H.degree():
+        degrees[d].add(node)
+        try:
+            H.lbound = min(H.lbound, d)  # Lower bound on min-degree.
+        except AttributeError:
+            # Python 2.X doesn't support `nonlocal`.
+            # Workaround: set as attribute of the smallest-last graph.
+            H.lbound = d
+
+    def find_min_degree():
+        for d in itertools.count(H.lbound):  # Save time starting here, not 0.
+            if d in degrees:
+                H.lbound = d  # This is clearly the lowest degree present.
+                return d
+
+    for _ in G:
+        # Pop a min-degree node and add it to the list.
+        min_degree = find_min_degree()
+        u = degrees[min_degree].pop()
+        if not degrees[min_degree]:  # Clean up the degree list.
+            del degrees[min_degree]
+        result.append(u)
+
+        # Update degrees of removed node's neighbors.
+        for v in H[u]:
+            degree = H.degree(v)
+            degrees[degree].remove(v)
+            if not degrees[degree]:  # Clean up the degree list.
+                del degrees[degree]
+            degrees[degree - 1].add(v)
+
+        # Finally, remove the node.
+        H.remove_node(u)
+        H.lbound -= 1  # This is a min node; decrement for tied neighbors.
+
+
+    return reversed(result)
```

Comment:
Is there a reason you use `list.append` with `reversed` instead of `deque.appendleft`? I think the latter will be more efficient, since no reverse will be necessary at the end.